File: leetcode/101-200/117-populating-next-right-pointers-in-each-node-ii/bench/connect.c
/* LeetCode #117 — C mirror (raw *Node), O(1)-space next-pointer population on an arbitrary tree.
 * Same algorithm and workload as connect.kara: each rep builds a ~500-node pseudo-random BST (fixed
 * shape, data-dependent value base), wires `next` with the dummy-head + tail threading, folds a
 * level hash. K = 16000. The metal floor: manual malloc/free, raw pointer chase. */
#include <stdio.h>
#include <stdlib.h>

#define MOD 1000000007LL

typedef struct Node {
    long long val;
    struct Node *left, *right, *next;
} Node;

static Node *new_node(long long v) {
    Node *n = (Node *)malloc(sizeof(Node));
    n->val = v;
    n->left = n->right = n->next = NULL;
    return n;
}

static Node *insert(Node *root, long long v) {
    if (root == NULL) return new_node(v);
    if (v < root->val) root->left = insert(root->left, v);
    else root->right = insert(root->right, v);
    return root;
}

static Node *build_bst(long long count, long long base) {
    Node *root = NULL;
    long long s = 88172645LL;
    for (long long i = 0; i < count; i++) {
        s = (s * 1103515245LL + 12345LL) % 2147483648LL;
        root = insert(root, (s % 100000LL) + base);
    }
    return root;
}

static void free_tree(Node *n) {
    if (!n) return;
    free_tree(n->left);
    free_tree(n->right);
    free(n);
}

static void connect(Node *root) {
    Node *leftmost = root;
    while (leftmost != NULL) {
        Node dummy;
        dummy.next = NULL;
        Node *tail = &dummy;
        Node *cur = leftmost;
        while (cur != NULL) {
            if (cur->left != NULL) {
                tail->next = cur->left;
                tail = cur->left;
            }
            if (cur->right != NULL) {
                tail->next = cur->right;
                tail = cur->right;
            }
            cur = cur->next;
        }
        leftmost = dummy.next;
    }
}

static long long level_hash(Node *root) {
    long long h = 1;
    Node *head = root;
    while (head != NULL) {
        Node *cur = head;
        while (cur != NULL) {
            h = (h * 131 + cur->val + 1) % MOD;
            cur = cur->next;
        }
        h = (h * 31 + 7) % MOD;
        Node *nh = NULL;
        Node *scan = head;
        while (scan != NULL) {
            if (scan->left != NULL) { nh = scan->left; break; }
            if (scan->right != NULL) { nh = scan->right; break; }
            scan = scan->next;
        }
        head = nh;
    }
    return h;
}

int main(void) {
    long long acc = 0;
    for (long long rep = 0; rep < 16000; rep++) {
        long long base = acc % 100;
        Node *root = build_bst(500, base);
        connect(root);
        long long h = level_hash(root);
        acc = (acc * 131 + h) % MOD;
        free_tree(root);
    }
    printf("%lld\n", acc);
    return 0;
}

File: leetcode/101-200/117-populating-next-right-pointers-in-each-node-ii/bench/connect.rs
// LeetCode #117 — Rust mirror, O(1)-space next-pointer population on an arbitrary tree.
// Rc<RefCell<Node>> — the *safe* interior-mutability peer to Kara's `shared` (RC); the `next`/`left`
// /`right` handles are rewritten in place, so RefCell is the honest safe-Rust counterpart (an unsafe
// raw-pointer Rust would match C). Same algorithm + workload as connect.kara: ~500-node BST (fixed
// shape, data-dependent value base), dummy-head + tail threading, K = 16000. Kara checks integer
// overflow by default, so the like-for-like row is `rustc -O -C overflow-checks=on`.
use std::cell::RefCell;
use std::rc::Rc;

const MOD: i64 = 1000000007;

struct Node {
    val: i64,
    left: Option<Rc<RefCell<Node>>>,
    right: Option<Rc<RefCell<Node>>>,
    next: Option<Rc<RefCell<Node>>>,
}

type Link = Option<Rc<RefCell<Node>>>;

fn new_node(v: i64) -> Rc<RefCell<Node>> {
    Rc::new(RefCell::new(Node {
        val: v,
        left: None,
        right: None,
        next: None,
    }))
}

fn insert(root: Link, v: i64) -> Link {
    match root {
        None => Some(new_node(v)),
        Some(n) => {
            if v < n.borrow().val {
                let l = n.borrow().left.clone();
                n.borrow_mut().left = insert(l, v);
            } else {
                let r = n.borrow().right.clone();
                n.borrow_mut().right = insert(r, v);
            }
            Some(n)
        }
    }
}

fn build_bst(count: i64, base: i64) -> Link {
    let mut root: Link = None;
    let mut s: i64 = 88172645;
    for _ in 0..count {
        s = (s * 1103515245 + 12345) % 2147483648;
        root = insert(root, (s % 100000) + base);
    }
    root
}

fn connect(root: &Link) {
    let mut leftmost = root.clone();
    while leftmost.is_some() {
        let dummy = new_node(0);
        let mut tail = dummy.clone();
        let mut cur = leftmost.clone();
        while let Some(c) = cur.clone() {
            let l = c.borrow().left.clone();
            if let Some(l) = l {
                tail.borrow_mut().next = Some(l.clone());
                tail = l;
            }
            let r = c.borrow().right.clone();
            if let Some(r) = r {
                tail.borrow_mut().next = Some(r.clone());
                tail = r;
            }
            cur = c.borrow().next.clone();
        }
        leftmost = dummy.borrow().next.clone();
    }
}

fn level_hash(root: &Link) -> i64 {
    let mut h: i64 = 1;
    let mut head = root.clone();
    while let Some(hd) = head.clone() {
        let mut cur = Some(hd.clone());
        while let Some(c) = cur.clone() {
            h = (h * 131 + c.borrow().val + 1) % MOD;
            cur = c.borrow().next.clone();
        }
        h = (h * 31 + 7) % MOD;
        let mut nh: Link = None;
        let mut scan = Some(hd.clone());
        while let Some(s) = scan.clone() {
            let l = s.borrow().left.clone();
            if l.is_some() {
                nh = l;
                break;
            }
            let r = s.borrow().right.clone();
            if r.is_some() {
                nh = r;
                break;
            }
            scan = s.borrow().next.clone();
        }
        head = nh;
    }
    h
}

fn main() {
    let mut acc: i64 = 0;
    for _ in 0..16000 {
        let base = acc % 100;
        let root = build_bst(500, base);
        connect(&root);
        let h = level_hash(&root);
        acc = (acc * 131 + h) % MOD;
    }
    println!("{}", acc);
}
